Automatic Image Type Detection in WatermarkService

KrishAnand369 opened the pull request:
🔍 Problem

Currently, when applying watermarks, users must manually specify an ImageType (e.g., PNG, JPEG).
This has several drawbacks:

•Users must know the image format beforehand.

•Errors occur if the wrong type is specified.

•The API feels less intuitive for common use cases.

✅ Solution

This PR introduces automatic image type detection in WatermarkService.

•Users can now call:

watermarkImage(File file)
watermarkImage(byte[] fileBytes)


without providing an ImageType.

•The library detects the format at runtime using ImageIO.

•If no suitable writer is found for the detected format, an UnsupportedOperationException is thrown.

•Existing methods that accept ImageType remain unchanged for backward compatibility.

🛠 Implementation

•Added watermarkImage(File file) and watermarkImage(byte[] fileBytes) overloads.

•Introduced ImageTypeDetector utility to identify image formats via ImageIO.

•Added runtime validation to ensure a writer exists before attempting ImageIO.write.

🧪 Testing

•Verified watermarking works with PNG, JPEG, and BMP test images without manually specifying the type.

•Added tests to confirm unsupported formats throw UnsupportedOperationException.

•All unit tests (mvn test) are passing.

📚 Notes

•No breaking changes: existing APIs using explicit ImageType continue to work.

•Improves developer experience by making the API easier and safer to use.

KrishAnand369 committed:
Add automatic image type detection for watermarking

Introduced ImageTypeDetector utility to detect image types from files and byte arrays. Updated WatermarkService and DefaultWatermarkService to support watermarking images without explicitly specifying the image type, improving usability and robustness.

KrishAnand369 committed:
Reformat ImageTypeDetector for consistent indentation

Updated the indentation in ImageTypeDetector.java to use spaces instead of tabs for improved code consistency and readability. No functional changes were made. Removed extra blank lines in AutoDetectImageTypeTest.kt.

KrishAnand369 committed:
Merge branch 'master' into enhancement/auto-detect-image-type

OlegCheban commented:
Thanks for the contribution. You have missed a closing brace in DefaultWatermarkService class. Fix it please. Also you should remove the old api methods which use imageType attribute. And remove imageType passing in old tests.

KrishAnand369 committed:
Remove imageType overloads from watermarkImage methods

Deprecated watermarkImage methods that accepted an explicit ImageType parameter have been removed from WatermarkService and DefaultWatermarkService. All image watermarking now relies on automatic image type detection, simplifying the API and usage. Related test code has been updated to reflect this change.

KrishAnand369 committed:
Merge branch 'master' into enhancement/auto-detect-image-type

KrishAnand369 commented:
made the recomended changes can you have a look

OlegCheban commented:
> made the recomended changes can you have a look

There is a new compilation error. Use `mvn clean install` to check any possible errors. In your case you need to rename DefaultWatermarkImageBuilder to WatermarkImageBuilder  in watermarkImage methods because DefaultWatermarkImageBuilder  is old name.

KrishAnand369 committed:
Replace DefaultWatermarkImageBuilder with WatermarkImageBuilder

Updated the DefaultWatermarkService to use WatermarkImageBuilder instead of DefaultWatermarkImageBuilder for image watermarking. This change ensures consistency with the intended builder class for image processing.

KrishAnand369 commented:
> > made the recomended changes can you have a look
> 
> There is a new compilation error. Use `mvn clean install` to check any possible errors. In your case you need to rename DefaultWatermarkImageBuilder to WatermarkImageBuilder in watermarkImage methods because DefaultWatermarkImageBuilder is old name.

done the changes
